File: 20. Travelling Salesman Problem (TSP).py
# 20. Travelling Salesman Problem (TSP)
# Topic: TSP (bitmask DP)
# Description: Use bitmasking + memoization to compute the shortest Hamiltonian cycle starting at node 0.
# Input: n followed by n x n distance matrix
# Output: minimal tour cost or -1 if not possible

import sys
from functools import lru_cache

INF = 10**18

if __name__ == '__main__':
    data = sys.stdin.read().strip().split()
    if not data:
        sys.exit(0)
    it = iter(data)
    n = int(next(it))
    dist = [[int(next(it)) for _ in range(n)] for _ in range(n)]

    @lru_cache(maxsize=None)
    def tsp(pos: int, mask: int) -> int:
        if mask == (1 << n) - 1:
            return dist[pos][0] if dist[pos][0] != 0 else INF
        best = INF
        for nxt in range(n):
            if mask & (1 << nxt):
                continue
            if dist[pos][nxt] == 0:
                continue
            best = min(best, dist[pos][nxt] + tsp(nxt, mask | (1 << nxt)))
        return best

    ans = tsp(0, 1)
    print(-1 if ans >= INF else ans)

# Memoization process started at tsp decorated with lru_cache.
# Time Complexity: O(n^2 * 2^n)
# Space Complexity: O(n * 2^n)

File: 20. Travelling Salesman Problem (TSP).cpp
// 20. Travelling Salesman Problem (TSP)
// Topic: TSP (bitmask DP)
// Description: Given n and distance matrix dist, compute minimal Hamiltonian cycle cost starting at
// 0. Input: n followed by n x n matrix Output: minimal tour cost

#include <bits/stdc++.h>
using namespace std;
using ll = long long;
const ll INF = (1LL << 60);

ll tsp(int pos, int mask, const vector<vector<ll>>& dist, vector<vector<ll>>& memo)
{
    int n = dist.size();
    if (mask == (1 << n) - 1)
        return dist[pos][0] ? dist[pos][0] : INF;  // return to start
    if (memo[pos][mask] != -1)
        return memo[pos][mask];
    ll best = INF;
    for (int nxt = 0; nxt < n; ++nxt)
    {
        if (mask & (1 << nxt))
            continue;
        if (dist[pos][nxt] == 0)
            continue;  // assume 0 means no edge
        best = min(best, dist[pos][nxt] + tsp(nxt, mask | (1 << nxt), dist, memo));
    }
    memo[pos][mask] = best;
    return best;
}

int main()
{
    ios::sync_with_stdio(false);
    cin.tie(nullptr);
    int n;
    if (!(cin >> n))
        return 0;
    vector<vector<ll>> dist(n, vector<ll>(n));
    for (int i = 0; i < n; ++i)
        for (int j = 0; j < n; ++j) cin >> dist[i][j];
    vector<vector<ll>> memo(n, vector<ll>(1 << n, -1));
    ll ans = tsp(0, 1 << 0, dist, memo);
    if (ans >= INF)
        cout << -1 << '\n';
    else
        cout << ans << '\n';
}

// Memoization process started in tsp where memo[pos][mask] caches best cost from state (pos,mask).
// Time Complexity: O(n^2 * 2^n)
// Space Complexity: O(n * 2^n)
